### PR TITLE
more reliable ASC (Neurolucida) parser

### PR DESCRIPTION
**Pull request by asanin-epfl**
- Allow neurite token to start with lower case character
- Allow soma token with whitespace: `(Cell Body)`
- Allow neurite token to be represented as text for backward compatibility with old morphologies as mentioned in #300 .

### CODE REVIEW

File: src/readers/lex.cpp [hunk 1]
```diff
@@ -104,6 +104,13 @@ constexpr std::size_t operator+(Token type) {
     return static_cast<std::size_t>(type);
 }
 
+const std::map<std::string, Token> NeuriteStringTokenMap{{to_string(Token::AXON), Token::AXON},
```
Comment (asanin-epfl) [inline src/readers/lex.cpp:107]: This format is due to clang. I didn't want this mess.

File: src/readers/morphologyASC.cpp [hunk 2]
```diff
@@ -40,6 +40,18 @@ bool is_neurite_type(Token id) {
             id == Token::CELLBODY);
 }
 
+std::string as_token_text(const std::string& text) {
+    std::string upp_text = text;
```
Comment (asanin-epfl) [inline src/readers/morphologyASC.cpp:44]: I'm open to suggestions here as I'm not C++ expert. It does:
- makes text upper case
- removes any whitespace

File: src/readers/morphologyASC.cpp [hunk 3]
```diff
@@ -217,6 +225,12 @@ class NeurolucidaParser
        Parse the root sexp until finding the first sexp containing numbers
     **/
     Header parse_root_sexp_header() {
+        static const std::map<std::string, Token> NeuriteStringTokenMap{
+            {to_string(Token::AXON), Token::AXON},
```
Comment (mgeplf) [inline src/readers/morphologyASC.cpp:229]: Other than CellBody, do any other tokens have this same behavior?

File: src/readers/morphologyASC.cpp (hunk 2 above)
Comment (asanin-epfl) [reply src/readers/morphologyASC.cpp:44]: Nope, any. it's for making `Cell Body` into `CellBody`.

Comment (mgeplf) [reply src/readers/morphologyASC.cpp:44]: Thanks.

File: src/readers/morphologyASC.cpp (hunk 3 above)
Comment (mgeplf) [reply src/readers/morphologyASC.cpp:229]: I believe that the old style was only used for CellBody/Soma; I have reduced the scope of the PR to that.

Comment (asanin-epfl) [reply src/readers/morphologyASC.cpp:229]: Should I update the PR with that in mind?

File: src/readers/morphologyASC.cpp [hunk 4]
```diff
@@ -217,6 +225,12 @@ class NeurolucidaParser
        Parse the root sexp until finding the first sexp containing numbers
     **/
     Header parse_root_sexp_header() {
+        static const std::map<std::string, Token> NeuriteStringTokenMap{
```
Comment (tomdele) [inline src/readers/morphologyASC.cpp:228]: for this usecase where we don t need stable iterators, don t care about ordering and having string as keys on a small map isn't an unordered_map a better choice ?

Comment (mgeplf) [reply src/readers/morphologyASC.cpp:228]: map is gone in the latest version

**Comment by asanin-epfl** [comment]
Should I do anything with coverage? What is your plan here @mgeplf ?

**Review by mgeplf — APPROVED** [review]

**Pull request merged**